manticore/core/cpu/disasm.py

```diff
@@ -0,0 +1,245 @@
+        :param pc: program counter
+        """
+        return next(self.disasm.disasm(code, pc))
+
+class BinjaILDisasm(Disasm):
+
+    def __init__(self, view):
+        self.view = view
+        # dictionary with llil for each function. This will be consumed
+        # using an iterator, so that we don't repeat ourselves whenever
+        # we ask for the next IL
+        self.func_llil = {}
+        # offset to account for section vs segment view of the binary
+        self.entry_point_diff = None
+        # current function
+        self.current_func = None
+        # current LowLevelILFunction
+        self.current_llil_func = None
+        # current pc
+        self.current_pc = None
+        # queue of il instructions at current PC. If we get called again from
+        # the same PC, we will pop from the queue
+        self.il_queue = []
+        # current il
+        self.disasm_il = None
+        # real size of the instruction
+        self.disasm_insn_size = None
+        # Binja LLIL size (size of operands)
+        self.insn_size = None
+
+        # for all UNIMPL insn and other hard times
+        # FIXME generalize for other archs
+        self.fallback_disasm = CapstoneDisasm(cs.CS_ARCH_X86, cs.CS_MODE_64)
+
+        super(BinjaILDisasm, self).__init__(view)
+
+    def _fix_addr(self, addr):
+        # FIXME how to deal with discrepancies of binja vs real program
+        # entry point addresses? We need to lookup the symbols and
+        # we should make sure all offsets are appropriate
+
+        if self.entry_point_diff is None:
+            # assume that the first time we are called, this is the entry point
+            # self.entry_point_diff = addr - self.view.entry_point
+            self.entry_point_diff = 0
+
+        return addr - self.entry_point_diff
+
+    def _pop_from_il_queue(self, code, pc):
+        # queue contains tuples of the form (idx, il_insn)
+        if self.il_queue != []:
+            # if we have multiple ils for the same pc
+            # pop from the front of the queue
+            if self.il_queue[0][1].address == pc:
+                return self.il_queue.pop(0)[1]
+            else:
+                # somehow we have a queue but we are now at a different pc,
+                # clear the queue (we might be here because of a CALL etc.)
+                # FIXME assert that this is legitimate
+                del self.il_queue[:]
+
+        from binaryninja import Architecture, LowLevelILFunction
+
+        func = LowLevelILFunction(self.view.arch)
+        func.current_address = pc
+        self.disasm_insn_size = (self.view.arch.
+                                 get_instruction_low_level_il(code, pc, func))
+        self.current_llil_func = func
+        self.il_queue = [(i, func[i]) for i in xrange(len(func))]
+        return self.il_queue.pop(0)[1]
+
+    def disassemble_instruction(self, code, pc):
+        """Get next instruction based on Capstone disassembler
```

Comment:
Inaccurate comment. Also, param docs below should have type annotations (i.e. `:param str code`)